volume_type_validation/volume_type_validation.py

```diff
@@ -0,0 +1,187 @@
+            container_nodes_taggings[container_nodes_taggings['name'] ==
+                                     "type"].copy()
+
+        if not type_taggings.empty:
+            # TODO For master nodes, it can happen
+            # that the nodes
+            # running etcd have io1 and the rest
+            # gp2. Lets ignore these for now, until we can
+            # identify nodes running etcd
+            non_master_nodes_ids = \
+                type_taggings[type_taggings['value'] !=
+                              "master"]['container_node_id']
+            self._find_invalid(
+                source_id,
+                container_nodes_group[container_nodes_group['id'].isin(non_master_nodes_ids)]
+            )
+        else:
+            self._find_invalid(source_id, container_nodes_group)
+
+    def _find_invalid(self, source_id, container_nodes_group):
+        # Select container nodes of 1 group, but only those deployed on Vm
+        vms_ids = container_nodes_group['lives_on_id']
+
+        # Select volume attachments of these container nodes
+        group_volume_attachments = \
+            self.volume_attachments[self.volume_attachments['vm_id'].isin(vms_ids)]
+        group_volume_attachments_ids = group_volume_attachments["volume_id"]
+
+        # Select volumes of these volume attachments
+        group_volumes = \
+            self.volumes[self.volumes['id'].isin(group_volume_attachments_ids)]
+
+        # Asserts
+        self._find_invalid_volume_types(source_id, group_volumes, group_volume_attachments, container_nodes_group)
+
+    def _find_invalid_volume_types(
+            self,
+            source_id,
+            group_volumes,
+            group_volume_attachments,
+            container_nodes_group
+    ):
+        # volume_type_id of the whole group of volumes should be the same, if not add error message
+        if len(group_volumes['volume_type_id'].unique()) > 1:
+            # group_volumes.groupby('volume_type_id').groups
+
+            # Find what is recommended volume type, it will be the one used by majority of nodes
+            volume_type_groups = group_volumes.groupby('volume_type_id').groups
+            volume_type_groups_sizes = \
+                [[key, len(group)] for key, group in volume_type_groups.items()]
+            recommended_volume_type, recommended_volume_type_size = \
+                max(volume_type_groups_sizes, key=operator.itemgetter(1))
+            same_sizes = [[key, value] for key, value in volume_type_groups_sizes if
+                          value == recommended_volume_type_size]
+
+            message = {}
+            message['cluster_name'] = self.sources[(self.sources['id'] == source_id)].name.values[0]
+            message['description'] = "Cluster contains one or more instances " \
+                                     "with different volume types. "
+            message['recommendations'] = []
+
+            if len(same_sizes) > 1:
+                message[
```

Comment:
This line end, should'n happen. I think this just slipped your attention.

Reply:
Yes, fixed that. Thanks.